lhotse/dataset/sampling.py

```diff
@@ -96,6 +98,9 @@ def _maybe_init_distributed(self, world_size: Optional[int], rank: Optional[int]
                 # A different error - pass it on.
                 raise
 
+    def _set_sampler_len(self):
+        self._len = sum(1 for item in self)
```

Comment:
Unfortunately, I don't think this works as expected. If you look at `__iter__`, it is going to shuffle the cuts every time it's called, so after this loop, the order of the cuts will be different.

Also this is not unset anywhere, so it is going to be incorrect in subsequent epochs.

Reply:
Hm so the shuffling should not happen in-place I guess?

Reply:
@pzelasko is it necessary that the expression `list(sampler)` gives correct result in the case of samplers? The `list()` constructor behaves different from eg `[item for item in sampler]`. The former seems to call `sampler.__len__` asynchronously many times while the latter simply calls `__iter__` and `__next__` in sequential order as is expected. 
Under the current state of code, I am getting `list(sampler) == []` as can be seen from the test. Any idea how this could be surpassed?

Reply:
one option is to shuffle in a deterministic way, so that the same seed + epoch will always result in the same shuffling order (this actually makes sense and makes it possible to re-start training from a checkpoint mid-epoch in an exact way, as we can know how many batches to skip and from which batch to resume)

Reply:
Oh, I did not know that `list(x)` behaves this way. Anyway with fully deterministic shuffling that shouldn't be an issue anymore I think.

Reply:
The shuffling you describe is basically how things work now I think. The iter method uses the self.seed + epoch random seed for shuffling so the process should be deterministic. I tried to determine the length already in init, but still getting this weird behavior.

Reply:
I don't think that's true, because we re-shuffle the list multiple times, so the elements are permuted in the same manner (because of the seed + epoch), but their order does change (just deterministically).

Reply:
Sorry I dont follow. Maybe it i would understand if i see some code. How the current `__iter__` should change? Right now it looks like this
```
    def __iter__(self) -> 'CutSampler':
        """
        Prepare the dataset for iterating over a new epoch. Will shuffle the data if requested.
        """
        if self.shuffle:
            r = random.Random(self.seed + self.epoch)
            r.shuffle(self.data_source)
        self.current_idx = 0
        self._len = None
        return self

```

Reply:
OK take a look at the following example:
```python
>>> x = list(range(10))
>>> x
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> r = Random(42)
>>> r.shuffle(x)
>>> x
[7, 3, 2, 8, 5, 6, 9, 4, 0, 1]
>>> r = Random(42)
>>> r.shuffle(x)
>>> x
[4, 8, 2, 0, 6, 9, 1, 5, 7, 3]
```

If you call iter multiple time, the same happens with cut ids (even though the seed + epoch are constant). We might need to hold the copy of the original cut ID ordering and re-compute the shuffling each time iter is called using the original order.

Reply:
Ok I updated the code to remember the original ordering. A list of indices is remembered and shuffled instead of the cut_ids. The random permutation does not change unless the user specifies a different seed. Also the length gets calculated correctly (number of sampled batches == len(sampler)) and the `list()` constructor works too. Is this what you had on mind?

Reply:
Great! Yes that sounds right. I will review later or tomorrow. Thanks!